### PR TITLE
Server Stats

### PR DESCRIPTION
**Pull request by ah1053**
Current Server Stats 

[live demo](http://autoclusteredgemultiplay-test.montreal-pitfield.telus.mobiledgex.net:7776/)

Screen Shot 2021-05-03 at 9 24 03 AM

### CODE REVIEW

**Review by brucearmstrong — APPROVED** [review]
Nice, clean UI. I like the links to the docs and Discord.

File: util/util.js [hunk 1]
```diff
@@ -55,7 +55,7 @@ function createRoom(lobby, playerId, playerName, playerAvatar, maxPlayersPerRoom
     if (connection === undefined) {
         connection.send(new events.NotificationEvent('create-room-faliure').convertToJSONString())
         console.log('cannot find player connection')
-        return
+        return undefined
```
Comment (lgarner) [inline util/util.js:58]: Hm, okay. Undefined is not writable (anymore).

**Review by lgarner — APPROVED** [review]
ok.

File: util/util.js [hunk 2]
```diff
@@ -138,12 +144,32 @@ function exitRoom(lobby, roomId, playerId) {
...
+    stats.membersInFullRooms = 0
+    if (lobby.availableRooms.size > 0) {
+        lobby.availableRooms.forEach((value) => { stats.membersInAvailableRooms += value.roomMembers.length })
+    }
+    if (lobby.fullRooms.size > 0) {
+        lobby.fullRooms.forEach((value) => { stats.membersInFullRooms += value.roomMembers.length })
+    }
+    stats.numOfConnectedClients = Object.keys(lobby.connectedClients).length
+    stats.membersStillInLobby = (stats.numOfConnectedClients - (stats.membersInAvailableRooms + stats.membersInFullRooms))
+    stats.lobbyRoomsOccupancy = ((1 - (lobby.rooms.size / lobby.MAX_ROOMS_PER_LOBBY)) * 100).toFixed(2) + '%'
```
Comment (franklin-huang-mobiledgex) [inline util/util.js:165]: Is lobby.rooms.size the number of occupied rooms? If so, I think occupancy should just be:

```
(lobby.rooms.size / lobby.MAX_ROOMS_PER_LOBBY) * 100).toFixed(2) + '%'
```

File: server.js [hunk 3]
```diff
@@ -86,14 +101,23 @@ wsServer.on('connection', function connection(ws, request) {
...
+                    if (roomId !== undefined) {
+                        wsStatsServer.emit('updateStats')
+                    }
                     break
                 case 'ExitRoom':
                     console.log('ExitRoom Request received from client %o', jsonObj)
-                    roomId = util.exitRoom(lobby, jsonObj.roomId, jsonObj.playerId)
+                    exitRoomSuccess = util.exitRoom(lobby, jsonObj.roomId, jsonObj.playerId)
+                    if (exitRoomSuccess === true) {
+                        wsStatsServer.emit('updateStats')
```
Comment (franklin-huang-mobiledgex) [inline server.js:119]: One thing that I think we probably want to do is aggregate some of these stats in memory for a short amount of time (10 seconds?) before pushing to your Websockets server. I'm thinking that if EdgeMultiplay gets super popular in the future, all of these emits directly to your server could overwhelm it. 

Also getLobbyStats has to recalculate all of the stats everytime an event happens, so if you aggregate these as you go, your server will have more bandwidth to do non-stats related things.

File: models/Room.js [hunk 4]
```diff
@@ -45,17 +45,23 @@ class Room {
...
+            this.udpConnections.delete(playerId)
+            this.roomMembers.forEach((roomMember, index, object) => {
+                if (roomMember.playerId === playerId) {
+                    object.splice(index, 1)
+                    updatedRoomMembers = object
+                }
+            })
+            for (counter = 0; counter < updatedRoomMembers.length; counter++) {
+                updatedRoomMembers[counter].playerIndex = counter
             }
```
Comment (franklin-huang-mobiledgex) [inline models/Room.js:58]: I still don't quite understand the point of playerIndex. Is this needed? If it is used for getting the "first" user for ownership, we can just grab any member from the updateRoomMembers

**Review by franklin-huang-mobiledgex — COMMENTED** [review]
This is great! See comments for some concerns about how often you send stats

File: util/util.js (hunk 2 above)
Comment (ah1053) [reply util/util.js]: Removed lobbyRoomsOccupancy for now, it's not used in the stats at least now.
Will add in the future with the correct calculation you mentioned.

File: server.js (hunk 3 above)
Comment (ah1053) [reply server.js:119]: Thanks for the suggestion, I agree will add those stats aggregation in the future.

File: models/Room.js (hunk 4 above)
Comment (ah1053) [reply models/Room.js:58]: As we discussed before, we hardcode playerIndex = 0 in multiple sample apps. Will remove it to a random number within the roomMembers.length range in the next release.
Thanks for the suggestion.

**Pull request merged**